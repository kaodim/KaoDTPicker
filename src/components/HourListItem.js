import React from 'react'
import PropTypes from 'prop-types'
import moment from 'moment'

import SurchargeArrow from './SurchargeArrow'
import SurchargeArrowWhite from './SurchargeArrowWhite'
import RebateArrow from './RebateArrow'
import RebateArrowWhite from './RebateArrowWhite'

const HourListItem = ({ timeslots, label, onChangeTime, selectedTime }) => {
  const btnStyle = (date) => {
    return selectedTime === date ? 'kld-hli__timeslot-item-selected' : 'kld-hli__timeslot-item'
  }
  const surchargeTextStyle = (date) => {
    return selectedTime === date
      ? 'kld-hli__timeslot-surcharge-selected'
      : 'kld-hli__timeslot-surcharge'
  }
  let surchargeArrowStyle = (date) => {
    return selectedTime === date ? (
      <SurchargeArrowWhite />
    ) : (
      <SurchargeArrow className="kld-daypicker__date-surcharge-icon" />
    )
  }
  let rebateArrowStyle = (date) => {
    return selectedTime === date ? (
      <RebateArrowWhite />
    ) : (
      <RebateArrow className="kld-daypicker__date-surcharge-icon" />
    )
  }
  return (
    <div className="kld-hli">
      <span className="kld-hli__zone">{label}</span>
      <div className="kld-hli__timeslot">
        {timeslots.map((tsItem, index) => {
          const date = tsItem.value
          const localTotalPrice = tsItem.localized_total_price
          const showPrice = tsItem.show_price
          const isRebatable = tsItem.rebatable
          const isSurchargable = tsItem.surchargable
          return (
            <button
              className={btnStyle(date)}
              disabled={!tsItem.available}
              key={index}
              onClick={() =>
                onChangeTime(date, localTotalPrice, showPrice, isRebatable, isSurchargable)
              }>
              <span>{moment(date).format('hh:mmA')}</span>
              {tsItem.show_price && (
                <div>
                  {tsItem.surchargable && surchargeArrowStyle(date)}
                  {tsItem.rebatable && rebateArrowStyle(date)}
                  <span className={surchargeTextStyle(date)}>{tsItem.human_readable_price}</span>
                </div>
              )}
            </button>
          )
        })}
      </div>
    </div>
  )
}

HourListItem.defaultProps = {
  label: '',
  onChangeTime: () => {},
  timeslots: [],
  selectedTime: ''
}

HourListItem.propTypes = {
  label: PropTypes.string,
  onChangeTime: PropTypes.func,
  timeslots: PropTypes.array,
  selectedTime: PropTypes.string
}

export default HourListItem
