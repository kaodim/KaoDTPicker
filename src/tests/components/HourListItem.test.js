import React from 'react'
import { shallow } from 'enzyme'
import HourListItem from '../../components/HourListItem'
import calTimeslots from '../fixtures/calTimeslots'

let wrapper
const zone = 'MORNING'
const date1 = '2020-01-17T07:00:00.000+08:00'
const date2 = '2020-01-17T08:00:00.000+08:00'

beforeEach(() => {
  wrapper = shallow(<HourListItem />)
})

test('should render HourListItem correctly', () => {
  expect(wrapper).toMatchSnapshot()
})

describe('should render HourListItem according to user interactions', () => {
  let spliceTimeslot = calTimeslots.splice(0, 1)
  beforeEach(() => {
    wrapper = shallow(<HourListItem label={zone} timeslots={spliceTimeslot} />)
  })

  test('should set selected class if props selectedtime is same', () => {
    wrapper.setProps({
      selectedTime: date1
    })
    expect(wrapper).toMatchSnapshot()
  })

  test('should set default class if props selectedtime is different', () => {
    wrapper.setProps({
      selectedTime: date2
    })
    expect(wrapper).toMatchSnapshot()
  })

  test('should call onChangeTime HourListItem', () => {
    const onChangeTimeSpy = jest.fn()
    wrapper.setProps({
      onChangeTime: onChangeTimeSpy(date1)
    })
    expect(onChangeTimeSpy).toHaveBeenLastCalledWith(date1)
  })

  test.skip('should call onClick when clicking on timeslot', () => {
    const onClickSpy = jest.fn()
    wrapper = shallow(
      <HourListItem onChangeTime={onClickSpy} label={zone} timeslots={spliceTimeslot} />
    )
    wrapper.find('button').simulate('click')
    expect(onClickSpy).toHaveBeenLastCalledWith(spliceTimeslot[0].value)
    expect(wrapper).toMatchSnapshot()
  })

  test('should have default onChangeTime', () => {
    expect(wrapper.instance().props.onChangeTime).toBeDefined()
  })
})
