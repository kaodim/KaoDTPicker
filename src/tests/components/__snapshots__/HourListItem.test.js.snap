// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render HourListItem according to user interactions should call onClick when clicking on timeslot 1`] = `
<div
  className="kld-hli"
>
  <span
    className="kld-hli__zone"
  >
    MORNING
  </span>
  <div
    className="kld-hli__timeslot"
  >
    <button
      className="kld-hli__timeslot-item"
      disabled={false}
      key="0"
      onClick={[Function]}
    >
      07:00AM
    </button>
  </div>
</div>
`;

exports[`should render HourListItem according to user interactions should set default class if props selectedtime is different 1`] = `
<div
  className="kld-hli"
>
  <span
    className="kld-hli__zone"
  >
    MORNING
  </span>
  <div
    className="kld-hli__timeslot"
  >
    <button
      className="kld-hli__timeslot-item"
      disabled={false}
      key="0"
      onClick={[Function]}
    >
      07:00AM
    </button>
  </div>
</div>
`;

exports[`should render HourListItem according to user interactions should set selected class if props selectedtime is same 1`] = `
<div
  className="kld-hli"
>
  <span
    className="kld-hli__zone"
  >
    MORNING
  </span>
  <div
    className="kld-hli__timeslot"
  >
    <button
      className="kld-hli__timeslot-item-selected"
      disabled={false}
      key="0"
      onClick={[Function]}
    >
      07:00AM
    </button>
  </div>
</div>
`;

exports[`should render HourListItem correctly 1`] = `
<div
  className="kld-hli"
>
  <span
    className="kld-hli__zone"
  />
  <div
    className="kld-hli__timeslot"
  />
</div>
`;
